In [117]:
import pandas as pd

# Load the datasets
train_features = pd.read_csv('training_set_features.csv')
train_labels = pd.read_csv('training_set_labels.csv')
test_features = pd.read_csv('test_set_features.csv')


# Display the first few rows of each dataset
print(train_features.head())
print(train_labels.head())
print(test_features.head())

# Merge the training features with the labels
data = pd.merge(train_features, train_labels, on='respondent_id')
data 
# Check for missing values
print(data.isnull().sum())


   respondent_id  xyz_concern  xyz_knowledge  behavioral_antiviral_meds  \
0              0          1.0            0.0                        0.0   
1              1          3.0            2.0                        0.0   
2              2          1.0            1.0                        0.0   
3              3          1.0            1.0                        0.0   
4              4          2.0            1.0                        0.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   0.0                   0.0                    0.0   
1                   1.0                   0.0                    1.0   
2                   1.0                   0.0                    0.0   
3                   1.0                   0.0                    1.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          0.0                      1.0  

In [118]:
data= data.drop(['employment_industry','employment_occupation','marital_status','rent_or_own','employment_status','hhs_geo_region','census_msa','household_adults','household_children','employment_industry','employment_occupation','health_insurance','income_poverty'
],axis=1)
test_features=test_features.drop(['employment_industry','employment_occupation','marital_status','rent_or_own','employment_status','hhs_geo_region','census_msa','household_adults','household_children','employment_industry','employment_occupation','health_insurance','income_poverty'
],axis=1)

In [119]:
print(data.isnull().sum())

respondent_id                     0
xyz_concern                      92
xyz_knowledge                   116
behavioral_antiviral_meds        71
behavioral_avoidance            208
behavioral_face_mask             19
behavioral_wash_hands            42
behavioral_large_gatherings      87
behavioral_outside_home          82
behavioral_touch_face           128
doctor_recc_xyz                2160
doctor_recc_seasonal           2160
chronic_med_condition           971
child_under_6_months            820
health_worker                   804
opinion_xyz_vacc_effective      391
opinion_xyz_risk                388
opinion_xyz_sick_from_vacc      395
opinion_seas_vacc_effective     462
opinion_seas_risk               514
opinion_seas_sick_from_vacc     537
age_group                         0
education                      1407
race                              0
sex                               0
xyz_vaccine                       0
seasonal_vaccine                  0
dtype: int64


In [120]:
data.shape

(26707, 27)

In [121]:

data['opinion_xyz_vacc_effective'] = data['opinion_xyz_vacc_effective'].fillna(data.opinion_xyz_vacc_effective.median())
data['opinion_xyz_risk'] = data['opinion_xyz_risk'].fillna(data.opinion_xyz_risk.median())
data['opinion_xyz_sick_from_vacc'] = data['opinion_xyz_sick_from_vacc'].fillna(data.opinion_xyz_sick_from_vacc.median())
data['opinion_seas_vacc_effective'] = data['opinion_seas_vacc_effective'].fillna(data.opinion_seas_vacc_effective.median())
data['opinion_seas_risk'] = data['opinion_seas_risk'].fillna(data.opinion_seas_risk.median())
data['opinion_seas_sick_from_vacc'] = data['opinion_seas_sick_from_vacc'].fillna(test_features.opinion_seas_sick_from_vacc.median())
test_features['opinion_xyz_vacc_effective'] = test_features['opinion_xyz_vacc_effective'].fillna(test_features.opinion_xyz_vacc_effective.median())
test_features['opinion_xyz_risk'] = test_features['opinion_xyz_risk'].fillna(test_features.opinion_xyz_risk.median())
test_features['opinion_xyz_sick_from_vacc'] = test_features['opinion_xyz_sick_from_vacc'].fillna(test_features.opinion_xyz_sick_from_vacc.median())
test_features['opinion_seas_vacc_effective'] = test_features['opinion_seas_vacc_effective'].fillna(test_features.opinion_seas_vacc_effective.median())
test_features['opinion_seas_risk'] = test_features['opinion_seas_risk'].fillna(test_features.opinion_seas_risk.median())
test_features['opinion_seas_sick_from_vacc'] = test_features['opinion_seas_sick_from_vacc'].fillna(test_features.opinion_seas_sick_from_vacc.median())

In [122]:
data=data.dropna()
test_features=test_features.dropna()

In [123]:
data.shape

(22624, 27)

In [124]:
df_dummies = pd.get_dummies(data, columns=['sex','race'], drop_first=True)
testdf_dummies = pd.get_dummies(test_features, columns=['sex','race'], drop_first=True)

In [125]:
df_dummies.education.unique()

array(['< 12 Years', '12 Years', 'Some College', 'College Graduate'],
      dtype=object)

In [126]:
category_mapping1 = {'< 12 Years': 0, '12 Years': 1, 'Some College': 2,'College Graduate':3}
df_dummies['education'] = df_dummies['education'].map(category_mapping1)
testdf_dummies['education'] = testdf_dummies['education'].map(category_mapping1)

In [127]:
df_dummies.age_group.unique()

array(['55 - 64 Years', '35 - 44 Years', '65+ Years', '45 - 54 Years',
       '18 - 34 Years'], dtype=object)

In [128]:
category_mapping2 = {'18 - 34 Years': 0, '35 - 44 Years': 1, '45 - 54 Years': 2,'55 - 64 Years':3 ,'65+ Years':4}
df_dummies['age_group'] = df_dummies['age_group'].map(category_mapping2)
testdf_dummies['age_group'] = testdf_dummies['age_group'].map(category_mapping2)

In [129]:
df_dummies['sex_Male']=df_dummies['sex_Male'].astype(int)
df_dummies['race_Hispanic']=df_dummies['race_Hispanic'].astype(int)
df_dummies['race_White']=df_dummies['race_White'].astype(int)
df_dummies['race_Other or Multiple']=df_dummies['race_Other or Multiple'].astype(int)
testdf_dummies['sex_Male']=testdf_dummies['sex_Male'].astype(int)
testdf_dummies['race_Hispanic']=testdf_dummies['race_Hispanic'].astype(int)
testdf_dummies['race_White']=testdf_dummies['race_White'].astype(int)
testdf_dummies['race_Other or Multiple']=testdf_dummies['race_Other or Multiple'].astype(int)

In [130]:
X_train=df_dummies.drop(['xyz_vaccine','seasonal_vaccine','respondent_id'], axis=1)
y_train_xyz=df_dummies['xyz_vaccine']
y_train_seasonal=df_dummies['seasonal_vaccine']
X_test=testdf_dummies.drop(['respondent_id'], axis=1)

In [131]:
from sklearn.linear_model import LogisticRegression
model_xyz=LogisticRegression()
model_seasonal=LogisticRegression()

In [132]:
model_xyz.fit(X_train,y_train_xyz)
model_seasonal.fit(X_train,y_train_seasonal)

C:\Users\khush\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [133]:
y_pred_xyz = model_xyz.predict_proba(X_test)[:, 1]
y_pred_seasonal = model_seasonal.predict_proba(X_test)[:, 1]

In [134]:
test_predictions = pd.DataFrame({
    'respondent_id': test_features['respondent_id'],
    'xyz_vaccine': y_pred_xyz,
    'seasonal_vaccine': y_pred_seasonal
})


In [135]:
test_predictions.to_csv('Khushal_Sethi_test_predictions.csv', index=False)

## Calculating ROC AUC Score using Training Data for both vaccines

In [136]:
from sklearn.model_selection import train_test_split
X_TRAIN, X_TEST, Y_TRAIN_xyz, Y_TEST_xyz, Y_TRAIN_seasonal, Y_TEST_seasonal= train_test_split(X_train,y_train_xyz,y_train_seasonal,test_size=0.2) 

In [137]:

model_xyz.fit(X_TRAIN,Y_TRAIN_xyz)
model_seasonal.fit(X_TRAIN,Y_TRAIN_seasonal)

LogisticRegression()

In [138]:

from sklearn.metrics import roc_auc_score
roc_auc_xyz = roc_auc_score(Y_TEST_xyz, model_xyz.predict_proba(X_TEST)[:, 1],average="macro")
roc_auc_seasonal = roc_auc_score(Y_TEST_seasonal, model_seasonal.predict_proba(X_TEST)[:, 1],average="macro")

In [139]:
print(f"ROC AUC for xyz_vaccine (validation): {roc_auc_xyz:.4f}")
print(f"ROC AUC for seasonal_vaccine (validation): {roc_auc_seasonal:.4f}")

ROC AUC for xyz_vaccine (validation): 0.8468
ROC AUC for seasonal_vaccine (validation): 0.8509


## So we have an ROC AUC Score of around 0.85